WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/domainregistration/DomainRegistrationSource.kt

```diff
@@ -74,25 +75,32 @@ class DomainRegistrationSource @Inject constructor(
         siteLocalId: Int,
         selectedSite: SiteModel
     ) {
-        continuation?.cancel()
-        continuation = null
-        coroutineScope.launch(bgDispatcher) {
-            try {
-                val event = suspendCancellableCoroutine<OnPlansFetched> { cancellableContinuation ->
-                    continuation = cancellableContinuation
-                    fetchPlans(selectedSite)
-                }
-                continuation = null
-                if (event.isError) {
-                    val message = "An error occurred while fetching plans : " + event.error.message
-                    appLogWrapper.e(DOMAIN_REGISTRATION, message)
+        if (continuations[siteLocalId] == null) {
+            coroutineScope.launch(bgDispatcher) {
+                try {
+                    val event = suspendCancellableCoroutine<OnPlansFetched> { cancellableContinuation ->
+                        continuations[siteLocalId] = cancellableContinuation
+                        fetchPlans(selectedSite)
+                    }
+                    when {
+                        event.isError -> {
+                            val message = "An error occurred while fetching plans : " + event.error.message
+                            appLogWrapper.e(DOMAIN_REGISTRATION, message)
+                            postState(DomainCreditAvailable(false))
+                        }
+                        siteLocalId == event.site.id -> {
+                            postState(DomainCreditAvailable(isDomainCreditAvailable(event.plans)))
+                        }
+                        else -> {
+                            postState(DomainCreditAvailable(false))
+                        }
+                    }
+                } catch (e: CancellationException) {
                     postState(DomainCreditAvailable(false))
-                } else if (siteLocalId == event.site.id) {
-                    postState(DomainCreditAvailable(isDomainCreditAvailable(event.plans)))
                 }
-            } catch (e: CancellationException) {
-                postState(DomainCreditAvailable(false))
             }
+        } else {
+            appLogWrapper.d(DOMAIN_REGISTRATION, "A request is already running for $siteLocalId")
         }
     }
 
```

Comment:
Suggestion (💡): Consider moving the `init { ... }` block and its accompanying `clear()` function to the top (below the fields) so that to increase visibility of the `dispatcher` logic for this source.

PS: If you like that idea, I suggestion doing the same for:
- `ScanAndBackupSource` and its `clear()` function.
- `SelectedSiteSource` and its `init { ... }` block and `clear()` function.

Reply:
You are correct, I could go back and move the logic around, but I assumed the writer followed their own pattern  when it was written ... chicken and egg here. Will someone come after us and prefer things to be moved around to what they think is better organization? 😜  

With that said, I'll be move the logic within this class, but keep the other two sources out of this PR; they are unrelated to the context. Thanks.

Reply:
In 479221c

Reply:
👋 @zwarm and thanks!

As far as the `chicken and egg` situation, and having other moved things around, especially in relation to `init { ... }` block I know we are doing the right thing as this need to be on top. The `clear()` function is debatable, I agree. I would just keep it close to the `init { ... }` block since the are grouped together already anyway.